### PR TITLE
[pkg] add recover and print cpu/mem/disk infos if crashed

### PR DESCRIPTION
**Pull request by huof6829**

### CODE REVIEW

**Comment by Liuhaai** [comment]
I think we can just one ` recovery()` [here](https://github.com/iotexproject/iotex-core/blob/master/server/main.go#L75)

File: pkg/recovery/recovery_test.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package recovery
+
+import (
+	"errors"
+	"testing"
+)
+
+func TestRecovery(t *testing.T) {
```
Comment (Liuhaai) [inline pkg/recovery/recovery_test.go:8]: Could you test `Recovery()` on your local machine about empty pointer access error, divide by zero error, and oom error? And Share the print results under this pr.

File: pkg/recovery/recovery.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		IOCounters map[string]disk.IOCountersStat `json:"io_counters"`
+		Partitions []disk.PartitionStat           `json:"partitions"`
+	}
+)
+
+// Recovery get the stack infos of current goroutine and the cpu/mem/disk current infos
+func Recovery() {
+	if r := recover(); r != nil {
+		log.S().Errorf("recover error :%v", r)
+		log.S().Errorf("%s", string(debug.Stack()))
```
Comment (CoderZhi) [inline pkg/recovery/recovery.go:49]: the `debug.Stack()` will be the stack of the panic code?

File: pkg/recovery/recovery.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	DiskInfo struct {
+		IOCounters map[string]disk.IOCountersStat `json:"io_counters"`
+		Partitions []disk.PartitionStat           `json:"partitions"`
+	}
+)
+
+// Recovery get the stack infos of current goroutine and the cpu/mem/disk current infos
+func Recovery() {
+	if r := recover(); r != nil {
+		log.S().Errorf("recover error :%v", r)
```
Comment (Liuhaai) [inline pkg/recovery/recovery.go:48]: Is it possible to use `pprof.WriteHeapProfile()` to log stack info when panic?

File: api/web3server.go [hunk 4]
```diff
@@ -136,6 +137,7 @@ func NewWeb3Server(core CoreService, httpPort int, cacheURL string, queryLimit u
 // Start starts the API server
 func (svr *Web3Server) Start(_ context.Context) error {
 	go func() {
+		defer recovery.Recovery()
```
Comment (Liuhaai) [inline api/web3server.go:140]: defer recovery.CrashLog()

File: pkg/recovery/recovery_test.go (hunk 1 above)
Comment (huof6829) [reply pkg/recovery/recovery_test.go:8]: ok

File: pkg/recovery/recovery.go (hunk 3 above)
Comment (huof6829) [reply pkg/recovery/recovery.go:48]: i can try it

File: pkg/recovery/recovery.go (hunk 2 above)
Comment (huof6829) [reply pkg/recovery/recovery.go:49]: yes

**Comment by huof6829** [comment]
> I think we can just one ` recovery()` [here](https://github.com/iotexproject/iotex-core/blob/master/server/main.go#L75). B/c it is the entrance of entire process.

File: api/web3server.go (hunk 4 above)
Comment (huof6829) [reply api/web3server.go]: ok

File: pkg/recovery/recovery_test.go (hunk 1 above)
Comment (huof6829) [reply pkg/recovery/recovery_test.go]: {"level":"error","ts":1647428975.904598,"caller":"server/main.go:109","msg":"recover error :runtime error: integer divide by zero","ioAddr":"io1xs77vnmnm4wwguma5js5cg9kzy750j3wegyyse"}
{"level":"info","ts":1647428975.9490592,"caller":"recovery/recovery.go:49","msg":"occur crash","ioAddr":"io1xs77vnmnm4wwguma5js5cg9kzy750j3wegyyse","stack":"goroutine 1 [running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x65\ngithub.com/iotexproject/iotex-core/pkg/recovery.CrashLog({0xc00025e120, 0x14ba9ed})\n\t/home/proj/iotex-core/pkg/recovery/recovery.go:49 +0x34\nmain.main.func1()\n\t/home/proj/iotex-core/server/main.go:111 +0x16a\npanic({0x12e3260, 0x25c4e70})\n\t/usr/local/go/src/runtime/panic.go:1038 +0x215\nmain.main()\n\t/home/proj/iotex-core/server/main.go:172 +0xb1b\n"}
{"level":"info","ts":1647428976.1808667,"caller":"recovery/recovery.go:72","msg":"occur crash","ioAddr":"io1xs77vnmnm4wwguma5js5cg9kzy750j3wegyyse","cpu":{"physical_count":1,"logical_count":1,"total_use_percent":[0],"per_use_percent":[0],"average_loads":{"load1":1.12,"load5":0.69,"load15":0.33},"running_times":[{"cpu":"cpu-total","user":964.35,"system":1331.51,"idle":47422.9,"nice":0,"iowait":151.13,"irq":0,"softirq":76.75,"steal":0,"guest":0,"guestNice":0},{"cpu":"cpu0","user":964.35,"system":1331.51,"idle":47422.9,"nice":0,"iowait":151.13,"irq":0,"softirq":76.75,"steal":0,"guest":0,"guestNice":0}],"infos":[{"cpu":0,"vendorId":"GenuineIntel","family":"6","model":"61","stepping":4,"physicalId":"0","coreId":"0","cores":1,"modelName":"Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz","mhz":2699.076,"cacheSize":3072,"flags":["fpu","vme","de","pse","tsc","msr","pae","mce","cx8","apic","sep","mtrr","pge","mca","cmov","pat","pse36","clflush","mmx","fxsr","sse","sse2","ht","syscall","nx","lm","constant_tsc","rep_good","nopl","xtopology","nonstop_tsc","cpuid","pni","pclmulqdq","ssse3","fma","cx16","pcid","sse4_1","sse4_2","movbe","popcnt","aes","xsave","avx","f16c","rdrand","hypervisor","lahf_lm","abm","3dnowprefetch","fsgsbase","bmi1","smep","bmi2","erms","rdseed","adx","smap","arat"],"microcode":""}]}}
{"level":"info","ts":1647428976.2090278,"caller":"recovery/recovery.go:72","msg":"occur crash","ioAddr":"io1xs77vnmnm4wwguma5js5cg9kzy750j3wegyyse","memory":{"virtaul":{"total":1027473408,"available":675422208,"used":189865984,"usedPercent":18.478919505038906,"free":352972800,"active":311439360,"inactive":243658752,"wired":0,"laundry":0,"buffers":16986112,"cached":467648512,"writeBack":0,"dirty":270336,"writeBackTmp":0,"shared":7786496,"slab":80384000,"sreclaimable":49098752,"sunreclaim":31285248,"pageTables":7127040,"swapCached":2363392,"commitLimit":1050603520,"committedAS":3333832704,"highTotal":0,"highFree":0,"lowTotal":0,"lowFree":0,"swapTotal":536866816,"swapFree":94289920,"mapped":53309440,"vmallocTotal":35184372087808,"vmallocUsed":14397440,"vmallocChunk":0,"hugePagesTotal":0,"hugePagesFree":0,"hugePageSize":2097152},"swap":{"total":536866816,"used":442576896,"free":94289920,"usedPercent":82.43699979400478,"sin":555323392,"sout":3190513664,"pgIn":0,"pgOut":0,"pgFault":0,"pgMajFault":0}}}
{"level":"info","ts":1647428976.2420993,"caller":"recovery/recovery.go:72","msg":"occur crash","ioAddr":"io1xs77vnmnm4wwguma5js5cg9kzy750j3wegyyse","disk":{"io_counters":{"vda":{"readCount":830681,"mergedReadCount":213449,"writeCount":137863,"mergedWriteCount":723506,"readBytes":16022957056,"writeBytes":15299457024,"readTime":394537,"writeTime":690609,"iopsInProgress":0,"ioTime":801954,"weightedIO":1150671,"name":"vda","serialNumber":"","label":""},"vda1":{"readCount":830605,"mergedReadCount":213449,"writeCount":137541,"mergedWriteCount":723506,"readBytes":16020671488,"writeBytes":15299457024,"readTime":394524,"writeTime":687196,"iopsInProgress":0,"ioTime":799908,"weightedIO":1081720,"name":"vda1","serialNumber":"","label":""}},"partitions":[{"device":"/dev/vda1","mountpoint":"/etc/resolv.conf","fstype":"ext4","opts":["rw","relatime","bind"]},{"device":"/dev/vda1","mountpoint":"/etc/hostname","fstype":"ext4","opts":["rw","relatime","bind"]},{"device":"/dev/vda1","mountpoint":"/etc/hosts","fstype":"ext4","opts":["rw","relatime","bind"]}]}}

**Review by CoderZhi — APPROVED** [review]

File: server/main.go [hunk 5]
```diff
@@ -102,6 +105,16 @@ func main() {
 		glog.Fatalln("Cannot config global logger, use default one: ", zap.Error(err))
 	}
 
+	defer func() {
+		if r := recover(); r != nil {
+			log.S().Errorf("crashlog: %v", r)
```
Comment (Liuhaai) [inline server/main.go:110]: move into `recovery.CrashLog()`

Comment (huof6829) [reply server/main.go:110]: `recover()` can not  act if move into `recovery.CrashLog()`

Comment (Liuhaai) [reply server/main.go:110]:
```
if r := recover(); r != nil {
	recovery.CrashLog(r, cfg.Log,)
}
```

Comment (huof6829) [reply server/main.go]: ok, it's clearly

File: pkg/recovery/recovery.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		IOCounters map[string]disk.IOCountersStat `json:"io_counters"`
+		Partitions []disk.PartitionStat           `json:"partitions"`
+	}
+)
+
+// CrashLog write down the current memory and stack info and the cpu/mem/disk infos into log dir
+func CrashLog(r interface{}, cfgLog log.GlobalConfig) {
+	log.S().Errorf("crashlog: %v", r)
+	if cfgLog.StderrRedirectFile != nil {
+		writeHeapProfile(filepath.Join(filepath.Dir(*cfgLog.StderrRedirectFile),
```
Comment (Liuhaai) [inline pkg/recovery/recovery.go:51]: we should have a default dir if cfgLog.StderrRedirectFile is nil?

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**